source/common/upstream/cluster_manager_impl.h

```diff
@@ -252,6 +278,7 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     Event::Dispatcher& thread_local_dispatcher_;
     std::unordered_map<std::string, ClusterEntryPtr> thread_local_clusters_;
     std::unordered_map<HostConstSharedPtr, ConnPoolsContainer> host_http_conn_pool_map_;
+    std::unordered_map<HostConstSharedPtr, TcpConnectionsMap> host_tcp_conn_map_;
```

Comment:
Do we need to clear this also here? I would have to refresh my memory on some of the lifetime issues in this class. https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L664

Reply:
I don't think I can just clear it, because it holds an object that's registered as a connection callback. But I could iterate through and call close on all the connections, which results in the map getting cleared.

Alternately, I could add an unregisterCallbacks function on connection, which would make handling and reasoning about lifetimes at shutdown much easier.

Reply:
Yeah, I think either would be fine. I would just add a test to track this case and then figure out a solution.

Reply:
I wrote an integration test; it looks like the current shutdown order already works properly. But I added an assert that the map is empty in that destructor, so we should catch it if ordering changes.

Reply:
Sorry I should have been more clear. It's not server shutdown that I'm worried about (though that might be an issue) but CDS cluster removal. I'm almost certain this will be broken if a cluster gets removed while there are active TCP connections.

Reply:
Yes, I agree it will be broken. I'll sort this out.

Reply:
Summary of offline conversation: 

When a cluster is removed we run https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L823. 

Because the connection-map is owned by the cluster manager, not by the cluster, there shouldn't be any lifetime issues when a cluster is removed.

Reply:
Added test coverage anyways

Reply:
Thanks, sorry for the runaround, clearly I needed more comments on my own code! :)